### PR TITLE
[PLUGIN-115] Implementation of Datastage functions

### PR DESCRIPTION
**Pull request by nitinmotgi**
This PR includes a few changes
- [x] Reorganized documentation [[1](https://github.com/data-integrations/wrangler/pull/401/commits/a215f8840ea23485f1f6150aeee0b9f6a42219c7)].
- [x] [Added new date and time functions](https://github.com/data-integrations/wrangler/pull/401/commits/a6167c73d2f11bb26046781b6373a5922ab49122).
- [x] [Added new null functions](https://github.com/data-integrations/wrangler/pull/401/commits/2c049a60c16049fd6babf59f3d05b7c1266415d6).
- [x] [Added logical test functions](https://github.com/data-integrations/wrangler/pull/401/commits/7e377a03fc77ef1f1ba76ecb2a88eccfac11381a). 
- [x] [Added number functions](https://github.com/data-integrations/wrangler/pull/401/commits/76195a8cf6e1f034af158bc5f50b0f980c15ad6e).
- [x] [Improved json functions](https://github.com/data-integrations/wrangler/pull/401/commits/77cedf9301dfc0d10a720536da35b1ea7f17d0a5).
- [x] [Improved geo functions](https://github.com/data-integrations/wrangler/pull/401/commits/04af24b516b379dd07a33e56ec42f7bbf65af225).
- [x] [Fixing EL error handling and messaging](https://github.com/data-integrations/wrangler/pull/401/commits/8639ffda914326b6ac584478391f3b7eda7aa2e8).

### CODE REVIEW

File: wrangler-api/src/main/java/io/cdap/wrangler/api/DirectiveExecutionException.java [hunk 1]
```diff
@@ -37,11 +37,11 @@ public DirectiveExecutionException(Throwable e) {
   }
 
   public DirectiveExecutionException(String directiveName, String errorMessage) {
-    this(String.format("Error encountered while executing '%s' : %s", directiveName, errorMessage));
+    this(String.format("Error executing '%s'- %s", directiveName, errorMessage));
   }
 
   public DirectiveExecutionException(String directiveName, String errorMessage, Throwable e) {
-    this(String.format("Error encountered while executing '%s' : %s", directiveName, errorMessage), e);
+    this(String.format("Error executing '%s'- %s", directiveName, errorMessage), e);
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/DirectiveExecutionException.java:44]: should we keep error message and separator consistent within wrangler? In [this](https://github.com/data-integrations/wrangler/pull/407) PR `:` is the separator.

File: wrangler-api/src/main/java/io/cdap/wrangler/api/DirectiveParseException.java [hunk 2]
```diff
@@ -44,11 +44,11 @@ public DirectiveParseException(String message) {
   }
 
   public DirectiveParseException(String directiveName, String errorMessage) {
-    this(String.format("Error encountered while parsing '%s' : %s", directiveName, errorMessage));
+    this(String.format("Error encountered while parsing '%s' %s", directiveName, errorMessage));
   }
 
   public DirectiveParseException(String directiveName, String errorMessage, Throwable e) {
-    this(String.format("Error encountered while parsing '%s' : %s", directiveName, errorMessage), e);
+    this(String.format("Error encountered while parsing '%s' %s", directiveName, errorMessage), e);
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/DirectiveParseException.java:51]: same here

File: README.md [hunk 3]
```diff
@@ -1,195 +1,22 @@
-# Data Prep
```
Comment (CuriousVini) [inline README.md:1]: why are we removing readme?

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+import com.jayway.jsonpath.spi.mapper.GsonMappingProvider;
+
+/**
+ * Collection of useful expression functions made available in the context
+ * of an expression.
+ *
+ * set-column column <expression>
+ */
+public final class JsonFunctions {
+  public static final Configuration GSON_CONFIGURATION = Configuration
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java]: this can be private

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+   */
+  private JsonFunctions() {}
+
+  /**
+   * Parses a column or string to JSON. This is equivalent to <code>JSON.parse()</code>
+   *
+   * @param json string representation of json.
+   * @return parsed json else throws an exception.
+   */
+  public static JsonElement Parse(String json) {
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java:63]: method names should start with small letters. https://www.oracle.com/technetwork/java/codeconventions-135099.html

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    try {
+      JsonElement element = PARSER.parse(json);
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java:65]: can do `return PARSER.parse(json);`

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+      return element;
+    } catch (JsonSyntaxException e) {
+      return JsonNull.INSTANCE;
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java:68]: This silently returns null json element. should this log any error message?

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+  /**
+   * Checks if a json is valid.
+   *
+   * @param json to checked for validity.
+   * @return true if valid, false otherwise.
+   */
+  public static boolean IsValid(String json) {
+    try {
+      JsonElement element = PARSER.parse(json);
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java]: can do `PARSER.parse(json);`

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  /**
+   * Checks if a Json is {@code JsonNull}.
+   *
+   * @param element to be inspected for null.
+   * @return true if null, false otherwise.
+   */
+  public static boolean IsNull(JsonElement element) {
+    if (element != null && element.isJsonNull()) {
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java:94]: can be simplified `return element != null && element.isJsonNull();`

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  /**
+   * Checks if a Json is {@code JsonObject}.
+   *
+   * @param element to be inspected for object.
+   * @return true if object, false otherwise.
+   */
+  public static boolean IsObject(JsonElement element) {
+    if (element != null && element.isJsonObject()) {
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java]: can be simplified `return element != null && element.isJsonObject();`

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  /**
+   * Checks if a Json is {@code JsonArray}.
+   *
+   * @param element to be inspected for array.
+   * @return true if array, false otherwise.
+   */
+  public static boolean IsArray(JsonElement element) {
+    if (element != null && element.isJsonArray()) {
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java]: `return element != null && element.isJsonArray();`

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+   * @param element json to be inspected.
+   * @param path to be searched for in the element.
+   * @param paths other paths.
+   * @return A json element containing the results of all json paths.
+   */
+  public static JsonElement Select(JsonElement element, String path, String ...paths) {
+    DocumentContext context = JsonPath.using(GSON_CONFIGURATION).parse(element);
+    if (paths.length == 0) {
+      return context.read(path);
+    } else {
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java]: there is no need for else here

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+    if (element == null) {
+      return GSON.toJson(JsonNull.INSTANCE);
+    }
+    return GSON.toJson(element);
+  }
+
+  /**
+   * @return Number of elements in the array.
+   */
+  public static int ArrayLength(JsonArray array) {
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java:158]: If array can be null, the method should be marked as nullable

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+      array.add((JsonElement) context.read(path));
+      for (String p : paths) {
+        array.add((JsonElement) context.read(p));
+      }
+      return array;
+    }
+  }
+
+  /**
+   * This method converts a JavaScript value to a JSON string.
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java]: Is this javadoc correct? `JavaScript` value?

File: wrangler-core/src/main/java/io/cdap/functions/NumberFunctions.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+   */
+  public static double MantissaFromDecimal(BigDecimal value) {
+    int exponent = Math.getExponent(value.doubleValue());
+    return value.doubleValue() / Math.pow(2, exponent);
+  }
+
+  public static double MantissaFromDFloat(Number value) {
+    return 0d;
+  }
+}
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/functions/NumberFunctions.java:56]: nit: new line

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java [hunk 16]
```diff
@@ -117,6 +115,24 @@ public ELResult execute(ELContext context, boolean nullMissingFields) throws ELE
     }
   }
 
+  private String constructErrMessage(JexlException e) {
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java]: should this be used in this PR as well: https://github.com/data-integrations/wrangler/pull/407

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java [hunk 17]
```diff
@@ -117,6 +115,24 @@ public ELResult execute(ELContext context, boolean nullMissingFields) throws ELE
...
+    if (e == null) {
+      return String.format("Unable to determine the cause of error.");
+    }
+    JexlInfo info = e.getInfo();
+    Throwable cause = e.getCause();
+    StringBuilder sb = new StringBuilder("");
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java]: this can be `StringBuilder sb = new StringBuilder("");`

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java (hunk 16 above)
Comment (CuriousVini) [reply wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java]: Also where is this function used?

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java [hunk 18]
```diff
@@ -44,7 +41,8 @@
 /**
  * This class <code>EL</code> is a Expression Language Handler.
  */
-public final class EL {
+public final class
+EL {
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java]: is this intentional change?

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java [hunk 19]
```diff
@@ -129,31 +145,25 @@ public ELResult execute(ELContext context) throws ELException {
     public Map<String, Object> functions() {
       Map<String, Object> functions = new HashMap<>();
       functions.put(null, Global.class);
-      functions.put("date", Dates.class);
-      functions.put("json", JSON.class);
-      functions.put("math", Math.class);
-      functions.put("string", StringUtils.class);
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java:135]: do we not need this function? Is there any impact on directives being applied due to removal of these functions?

File: wrangler-core/src/test/java/io/cdap/directives/parser/JsPathTest.java [hunk 20]
```diff
@@ -70,16 +70,17 @@ public void testJSONFunctions() throws Exception {
...
-      "set-column s11 json:select(body, true, \"$.numbers\")",
-      "set-column s2 json:select(body, true, \"$.numbers\")",
-      "set-column s4 json:drop(body, \"numbers\", \"integer\", \"float\", \"aliases\", \"name\")",
-      "set-column s5 json:join(s11, \":\")"
+      "set-column body json:Parse(body)",
+      "set-column s0 json:Select(body, '$.name.fname', '$.name.lname')",
+      "set-column s1 json:Select(body, '$.name.fname')",
+      "set-column s11 json:Select(body, '$.numbers')",
+      "set-column s2 json:Select(body, '$.numbers')",
+      "set-column s6 json:ArrayLength(json:Select(body, '$.numbers'))",
```
Comment (CuriousVini) [inline wrangler-core/src/test/java/io/cdap/directives/parser/JsPathTest.java:78]: why are these changes needed? Does this mean whoever is using wrangler will have to reapply `set-column` with new syntax?

File: wrangler-core/src/test/java/io/cdap/functions/LogicalTest.java [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+    Assert.assertEquals(240L, rows.get(0).getValue("xor"));
+    Assert.assertEquals(352L, rows.get(0).getValue("compress"));
+    Assert.assertEquals("101100000", rows.get(0).getValue("expand"));
+    Assert.assertEquals(expected, rows.get(0).getValue("not1"));
+    Assert.assertEquals(expected, rows.get(0).getValue("not2"));
+    Assert.assertEquals(expected, rows.get(0).getValue("not3"));
+    Assert.assertEquals(494L, rows.get(0).getValue("bitset"));
+  }
+
+}
```
Comment (CuriousVini) [inline wrangler-core/src/test/java/io/cdap/functions/LogicalTest.java]: nit: new line

**Review by CuriousVini — COMMENTED** [review]
took a first pass. A few comments

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java (hunk 5 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java:63]: This is used by user within the expression e.g. set-column :col1 exp:{ json:Parse(foo) } in wrangler. @CuriousVini

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java (hunk 7 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java:68]: Not really, because this happens in parsing of data, it will fill the logs.

File: wrangler-api/src/main/java/io/cdap/wrangler/api/DirectiveExecutionException.java (hunk 1 above)
Comment (nitinmotgi) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/DirectiveExecutionException.java:44]: They are different. That's error row that's moved to error port.

File: wrangler-api/src/main/java/io/cdap/wrangler/api/DirectiveParseException.java (hunk 2 above)
Comment (nitinmotgi) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/DirectiveParseException.java:51]: They are different. That's error row that's moved to error port.

File: README.md (hunk 3 above)
Comment (nitinmotgi) [reply README.md:1]: We didn't it was years old, so cleaned it up.

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java (hunk 4 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java]: done

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java (hunk 6 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java:65]: Want to catch exception and return null instance as this will used within jexl

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java (hunk 8 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java]: done!

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java (hunk 10 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java]: done!

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java (hunk 14 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java]: fixed

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java (hunk 16 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java]: That's ErrorRecord which is not same as finding this error.

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java (hunk 18 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java]: I don't see that.

File: wrangler-core/src/test/java/io/cdap/directives/parser/JsPathTest.java (hunk 20 above)
Comment (nitinmotgi) [reply wrangler-core/src/test/java/io/cdap/directives/parser/JsPathTest.java:78]: They were broken, so I don't even know if anyone was using them.

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java (hunk 19 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java:135]: DIdn't remove, just changed the classes and separated them correctly.

**Comment by nitinmotgi** [comment]
@CuriousVini can you please take a look at it again

**Comment by albertshau** [comment]
This was broken up into smaller PRs and merged.
